single_kernel_mongo/events/ldap.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Handlers for ldap relations: ldap, ldap-certificates and ldap-peer."""
+
+from __future__ import annotations
+
+import logging
+from typing import TYPE_CHECKING
+
+from ops.framework import EventBase, EventSource, Object
+from ops.model import BlockedStatus, WaitingStatus
+
+from single_kernel_mongo.config.relations import ExternalRequirerRelations
+from single_kernel_mongo.exceptions import (
+    DeferrableError,
+    DeferrableFailedHookChecksError,
+    LDAPSNotEnabledError,
+    NonDeferrableFailedHookChecksError,
+    WaitingForLdapDataError,
+)
+from single_kernel_mongo.lib.charms.certificate_transfer_interface.v0.certificate_transfer import (
+    CertificateAvailableEvent,
+    CertificateRemovedEvent,
+)
+from single_kernel_mongo.lib.charms.glauth_k8s.v0.ldap import LdapReadyEvent, LdapUnavailableEvent
+from single_kernel_mongo.utils.event_helpers import defer_event_with_info_log
+
+if TYPE_CHECKING:
+    from single_kernel_mongo.abstract_charm import AbstractMongoCharm
+    from single_kernel_mongo.core.operator import OperatorProtocol
+
+logger = logging.getLogger(__name__)
+
+
+class RestartIfReadyEvent(EventBase):
+    """Handles the restart of the units if they are ready."""
+
+
+class LDAPEventHandler(Object):
+    """Event Handler for all ldap related relations."""
+
+    restart_if_ready_event = EventSource(RestartIfReadyEvent)
+
+    def __init__(self, dependent: OperatorProtocol):
+        self.dependent: OperatorProtocol = dependent
+        self.charm: AbstractMongoCharm = dependent.charm
+        self.manager = dependent.ldap_manager
+        self.relation_name = ExternalRequirerRelations.LDAP.value
+        self.cert_relation_name = ExternalRequirerRelations.LDAP_CERT.value
+
+        super().__init__(parent=dependent, key=self.relation_name)
+
+        self.framework.observe(self.manager.ldap_requirer.on.ldap_ready, self._on_ldap_ready)
+        self.framework.observe(
+            self.manager.ldap_requirer.on.ldap_unavailable, self._on_ldap_unavailable
+        )
+        self.framework.observe(
+            self.manager.certificate_transfer.on.certificate_available,
+            self._on_certificate_available,
+        )
+        self.framework.observe(
+            self.manager.certificate_transfer.on.certificate_removed, self._on_certificate_removed
+        )
+        self.framework.observe(self.restart_if_ready_event, self._on_restart_if_ready)
+
+        self.framework.observe(
+            self.charm.on["ldap-peers"].relation_changed, self._on_restart_if_ready
+        )
+
+    def _on_ldap_ready(self, event: LdapReadyEvent):
```

Comment:
And in all subsequent methods
```suggestion
    def _on_ldap_ready(self, event: LdapReadyEvent) -> None:
```